File: app/src/main/java/com/codelog/schyfts/CallSchedule.java
package com.codelog.schyfts;

import com.codelog.clogg.Logger;
import com.codelog.schyfts.api.APIException;
import com.codelog.schyfts.api.APIRequest;
import com.codelog.schyfts.api.CallData;
import com.codelog.schyfts.api.Doctor;
import com.codelog.schyfts.util.AlertFactory;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.util.Callback;
import javafx.util.StringConverter;

import java.io.IOException;
import java.net.URL;
import java.time.LocalDate;
import java.util.List;
import java.util.ResourceBundle;

public class CallSchedule implements Initializable {

    @FXML
    private Button btnSubmit;
    @FXML
    private TableView<CallData> tblCalls;
    @FXML
    private MenuItem mnuAdd;
    @FXML
    private MenuItem mnuDelete;
    @FXML
    private ProgressBar prgStatus;
    @FXML
    private ChoiceBox<Doctor> cmbDoctor;
    @FXML
    private ChoiceBox<Boolean> cmbCall;
    @FXML
    private DatePicker dpDate;

    List<Doctor> doctors;
    List<CallData> callData;
    boolean submitAction = false;

    public void refresh() {
        btnSubmit.setDisable(true);
        doctors = Doctor.getAllDoctors();
        prgStatus.setProgress(0.33);

        callData = CallData.getAllCallData();
        prgStatus.setProgress(0.67);

        tblCalls.getItems().clear();
        tblCalls.getItems().addAll(callData);

        btnSubmit.setDisable(false);

        cmbDoctor.getItems().clear();
        cmbDoctor.getItems().addAll(doctors);

        prgStatus.setProgress(1);
    }

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        tblCalls.setColumnResizePolicy(TableView.CONSTRAINED_RESIZE_POLICY);
        tblCalls.getColumns().clear();
        tblCalls.getItems().clear();

        TableColumn<CallData, Integer> clmDoctorId = new TableColumn<>("Doctor ID");
        clmDoctorId.setCellValueFactory(new PropertyValueFactory<>("doctorID"));

        TableColumn<CallData, String> clmName = new TableColumn<>("Name");
        clmName.setCellValueFactory(new PropertyValueFactory<>("name"));

        TableColumn<CallData, String> clmSurname = new TableColumn<>("Surname");
        clmSurname.setCellValueFactory(new PropertyValueFactory<>("surname"));

        TableColumn<CallData, LocalDate> clmDate = new TableColumn<>("Date");
        clmDate.setCellValueFactory(new PropertyValueFactory<>("date"));

        TableColumn<CallData, String> clmState = new TableColumn<>("Call Status");
        clmState.setCellValueFactory(new PropertyValueFactory<>("state"));

        tblCalls.getColumns().add(clmDoctorId);
        tblCalls.getColumns().add(clmName);
        tblCalls.getColumns().add(clmSurname);
        tblCalls.getColumns().add(clmDate);
        tblCalls.getColumns().add(clmState);

        mnuAdd.setOnAction(event -> {
            submitAction = true;
            btnSubmit.setText("Add");
        });

        AlertFactory.showAndWait("Refreshing call entries...");

        cmbDoctor.setConverter(new StringConverter<>() {
            @Override
            public String toString(Doctor object) {
                if (object == null)
                    return "";
                return "%s, %s".formatted(object.getSurname(), object.getName());
            }

            @Override
            public Doctor fromString(String string) {
                for (var d : doctors) {
                    var fullName = "%s, %s".formatted(d.getSurname(), d.getName());
                    if (fullName.equals(string))
                        return d;
                }
                return null;
            }
        });

        cmbCall.getItems().addAll(true, false);
        cmbCall.setConverter(new StringConverter<>() {
            @Override
            public String toString(Boolean object) {
                if (object == null)
                    return "CALL OFF";
                return object ? "CALL ON" : "CALL OFF";
            }

            @Override
            public Boolean fromString(String string) {
                return string.contains("ON");
            }
        });

        Thread refreshThread = new Thread(this::refresh);
        refreshThread.start();
    }

    public void btnSubmitClick(ActionEvent actionEvent) {

        if (cmbDoctor.getSelectionModel().isEmpty() || cmbCall.getSelectionModel().isEmpty()
                || dpDate.getValue() == null)
            AlertFactory.showAlert(Alert.AlertType.ERROR, "Please complete all fields!");

        if (submitAction) {
            APIRequest req = new APIRequest("addCall", true, "dID", "date", "value");
            var selectedDoctor = cmbDoctor.getSelectionModel().getSelectedItem();
            var dID = selectedDoctor.getId();
            var date = dpDate.getValue();
            var value = cmbCall.getValue();

            try {
                req.send(dID, date.toString().split("T")[0], value);
                AlertFactory.showAlert("Call added!");
            } catch (IOException | APIException e) {
                AlertFactory.showAlert(Alert.AlertType.ERROR, e.getMessage());
                Logger.getInstance().error("Error: Couldn't add call!");
                if (e instanceof APIException)
                    Logger.getInstance().error(((APIException)e).getApiResponse().toString(4));
                Logger.getInstance().exception(e);
            }

        } else {
            AlertFactory.showAndWait(Alert.AlertType.ERROR, "Function not implemented yet!");
        }

        cmbCall.getSelectionModel().clearSelection();
        cmbDoctor.getSelectionModel().clearSelection();
        dpDate.getEditor().clear();
        btnSubmit.setText("Submit");
        submitAction = false;
        Thread t = new Thread(this::refresh);
        t.start();
    }

    public void mnuDeleteClick(ActionEvent actionEvent) {
        if (tblCalls.getSelectionModel().isEmpty())
            return;

        try {
            APIRequest req = new APIRequest("deleteCall", true, "id");
            req.send(tblCalls.getSelectionModel().getSelectedItem().getId());
        } catch (IOException | APIException e) {
            AlertFactory.showAlert(Alert.AlertType.ERROR, "Couldn't delete call!");
            Logger.getInstance().exception(e);
            return;
        }

        AlertFactory.showAlert("Call deleted!");
        Thread t = new Thread(this::refresh);
        t.start();
    }
}

File: app/src/main/kotlin/com/codelog/schyfts/CallSchedule.kt
package com.codelog.schyfts

import com.codelog.schyfts.api.APIException
import com.codelog.schyfts.api.APIRequest
import com.codelog.schyfts.api.Doctor
import com.codelog.schyfts.util.AlertFactory
import com.codelog.schyfts.util.DialogFactory
import com.codelog.schyfts.util.DoctorStringConverter
import javafx.beans.InvalidationListener
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.control.*
import javafx.scene.control.cell.ChoiceBoxTableCell
import javafx.scene.control.cell.PropertyValueFactory
import javafx.scene.input.KeyCode
import javafx.stage.Stage
import javafx.util.Callback
import org.json.JSONArray
import org.json.JSONObject
import java.net.URL
import java.time.DayOfWeek
import java.time.LocalDate
import java.util.*
import kotlin.random.Random
import com.codelog.schyfts.util.KLoggerContext as Logger

data class CallEntry(val calls: Array<Doctor>, val dow: DayOfWeek) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as CallEntry

        if (!calls.contentEquals(other.calls)) return false
        if (dow != other.dow) return false

        return true
    }

    override fun hashCode(): Int {
        var result = calls.contentHashCode()
        result = 31 * result + dow.hashCode()
        return result
    }

    fun serialize(): JSONObject {
        val returnVal = JSONObject()
        val callsArray = JSONArray()

        for (doctor in calls) {
            callsArray.put(doctor.id)
        }

        returnVal.put("calls", callsArray)
        returnVal.put("dow", dow.ordinal)

        return returnVal
    }
}

fun Doctor.getFullName(reverse: Boolean = false): String =
    if (reverse)
        "$surname, $name"
    else
        "$name $surname"

class CallSchedule: Initializable {
    companion object {
        const val CALLS = 3
        var stage: Stage? = null

        private var instance: CallSchedule? = null

        fun registerEventListeners() {
            stage?.scene?.setOnKeyPressed {
                if (it.isAltDown && it.code == KeyCode.R) {
                    val rand = Random(System.nanoTime())

                    if (instance != null) {
                        val notNullInstance = instance ?: CallSchedule()
                        for (entry in notNullInstance.tblCalls.items) {
                            for (i in 0..2) {
                                val selection = rand.nextInt(0, notNullInstance.doctors.size)
                                entry.calls[i] = notNullInstance.doctors[selection]
                            }
                        }

                        notNullInstance.tblCalls.refresh()
                        notNullInstance.isSaved = false
                    }
                }
            }
        }
    }

    @FXML
    lateinit var tblCalls: TableView<CallEntry>
    lateinit var doctors: List<Doctor>

    var isSaved = false

    override fun initialize(location: URL?, resources: ResourceBundle?) {
        doctors = Doctor.getAllDoctors()
        val entries = ArrayList<CallEntry>()
        tblCalls.isEditable = true

        for (day in DayOfWeek.values())
            entries.add(CallEntry(Array(3) { Doctor(0, "", "", "", "") }, day))

        val clmDay: TableColumn<CallEntry, DayOfWeek> = TableColumn("Date")
        clmDay.cellValueFactory = PropertyValueFactory("dow")
        tblCalls.columns.add(clmDay)

        for (i in 1..CALLS) {
            val clmCall = TableColumn<CallEntry, Doctor>("Call $i")
            clmCall.cellValueFactory = Callback {
                return@Callback object: ObservableValue<Doctor> {
                    override fun addListener(listener: ChangeListener<in Doctor>?) {
                    }

                    override fun addListener(listener: InvalidationListener?) {
                    }

                    override fun removeListener(listener: InvalidationListener?) {
                    }

                    override fun removeListener(listener: ChangeListener<in Doctor>?) {
                    }

                    override fun getValue(): Doctor {
                        return it.value.calls[i - 1]
                    }
                }
            }

            val doctorsObservable: ObservableList<Doctor> = FXCollections.observableArrayList()
            for (d in doctors)
                doctorsObservable.add(d)

            val doctorsSorted = doctorsObservable.sorted { o1, o2 ->
                o1.getFullName(true).compareTo(o2.getFullName(true))
            }

            clmCall.cellFactory = ChoiceBoxTableCell.forTableColumn(DoctorStringConverter(), doctorsSorted)
            clmCall.isEditable = true

            clmCall.setOnEditCommit {
                it.rowValue.calls[i - 1] = it.newValue
                isSaved = false
            }

            tblCalls.columns.add(clmCall)
        }

        tblCalls.items.addAll(entries)

        instance = this


    }

    fun mnuSaveClick() {
        val dialog = DialogFactory.makeDatePickerDialog("Please select a date on which this week starts.");

        val result = dialog.showAndWait()
        if (result.isEmpty)
            return

        if (result.get().dayOfWeek != DayOfWeek.MONDAY) {
            AlertFactory.showAndWait(Alert.AlertType.ERROR, "That date is not a Monday!")
            return
        }

        val jsonData = JSONObject()
        jsonData.put("date", result.get().toString())
        val entryArray = JSONArray()
        for (entry in tblCalls.items) {
            entryArray.put(entry.serialize())
        }
        jsonData.put("entries", entryArray)

        val request = APIRequest("updateCallRegistry", true, "date", "entries")
        try {
            val response = request.send(result.get().toString(), entryArray)
            if (response.getString("status") == "ok") {
                AlertFactory.showAndWait("Done!")
                isSaved = true
            } else {
                throw APIException("Couldn't save to database!", response)
            }
        } catch (e: Exception) {
            Logger.error("Couldn't save call schedule!")
            Logger.exception(e)

            AlertFactory.showAndWait(Alert.AlertType.ERROR, "Couldn't save to database!");
        }
    }

    private fun savePrompt(): Boolean? {
        val dialog = Dialog<Boolean?>()
        dialog.title = "Save to database?"
        dialog.headerText = "The schedule has not been saved yet. Would you like to save it now?"
        dialog.dialogPane.buttonTypes.addAll(ButtonType.YES, ButtonType.NO, ButtonType.CANCEL)

        dialog.setResultConverter {
            when (it) {
                ButtonType.CANCEL -> null
                ButtonType.YES -> true
                ButtonType.NO -> false
                else -> null
            }
        }

        val result = dialog.showAndWait()

        return result.get()
    }

    fun mnuClearClick() {
        if (!isSaved) {
            val saveResponse = savePrompt() ?: return
            if (saveResponse)
                mnuSaveClick()
        }

        val entries = ArrayList<CallEntry>()
        for (day in DayOfWeek.values()) {
            entries.add(
                CallEntry(
                    Array(3) {
                        Doctor(0, "", "", "", "")
                    }, day
                )
            )
        }

        tblCalls.items.clear()
        tblCalls.items.addAll(entries)
        tblCalls.refresh()
        isSaved = false
    }

    fun mnuLoadClick() {
        if (!isSaved) {
            val saveResponse = savePrompt() ?: return
            if (saveResponse)
                mnuSaveClick()
        }

        val dialog = DialogFactory.makeDatePickerDialog("Please select the Monday of the week you want to load.")
        val result = dialog.showAndWait()

        if (result.isEmpty)
            return

        if (result.get().dayOfWeek != DayOfWeek.MONDAY) {
            AlertFactory.showAlert(Alert.AlertType.WARNING, "That date is not a Monday!");
            return;
        }

        val baseDate = result.get();
        val startDate = baseDate.toString().split("T")[0]
        val endDate = baseDate.plusDays(7).toString().split("T")[0]

        val req = APIRequest("getCallData", true, "start", "end")
        try {
            val jsonArray = req.send(startDate, endDate).getJSONArray("results")
            tblCalls.items.clear();

            val entries = ArrayList<CallEntry>(jsonArray.length() / 3 + 1)

            for (dow in DayOfWeek.values()) {
                val calls = Array(3) { Doctor(0, "", "", "", "") }
                for (i in 0 until jsonArray.length()) {
                    val json = jsonArray.getJSONObject(i)
                    val date = LocalDate.parse(json.getString("date").split("T")[0])
                    if (date.dayOfWeek == dow) {
                        val dID = json.getInt("dID")
                        val value = json.getInt("value")
                        val doctor = doctors.find { it.id == dID } ?: Doctor(0, "", "", "", "")
                        calls[value - 1] = doctor;
                    }
                }
                entries.add(CallEntry(calls, dow))
            }

            entries.sortWith(Comparator { o1, o2 -> return@Comparator o1.dow.ordinal - o2.dow.ordinal })
            tblCalls.items.addAll(entries);
            tblCalls.refresh()
        } catch (e: APIException) {
            Logger.error("Couldn't fetch Call Data!");
            Logger.exception(e);
        }
    }
}